tests/bdd/features/volume_group/set_replica/test_set_replica.py

```diff
@@ -0,0 +1,292 @@
+
+
+@pytest.fixture(autouse=True)
+@given("an io-engine cluster")
+def an_ioengine_cluster():
+    """an io-engine cluster."""
+    Deployer.start(
+        io_engines=3,
+    )
+    pytest.vol_id = ""
+    yield
+    Deployer.stop()
+
+
+@when("the 3 volumes are scaled down to 2")
+def the_3_volumes_are_scaled_down_to_2():
+    """the 3 volumes are scaled down to 2."""
+    scale_volume_group(2)
+
+
+@when("the 3 volumes are scaled down to 1")
+def the_3_volumes_are_scaled_down_to_1():
+    """the 3 volumes are scaled down to 1."""
+    try:
+        scale_volume_group(1)
+    except openapi.exceptions.ServiceException as e:
+        pytest.error_body = json.loads(e.body)
+
+
+@when("the 3 volumes are scaled up by one")
+def the_3_volumes_are_scaled_up_by_one():
+    """the 3 volumes are scaled up by one."""
+    scale_volume_group(2)
+
+
+@when("the volume is scaled up by one")
+def the_volume_is_scaled_up_by_one():
+    """the volume is scaled up by one."""
+    ApiClient.volumes_api().put_volume_replica_count(VG_VOLUME_UUID_1, 2)
+
+
+@when(
+    parsers.parse(
+        "volume belonging to a volume group with one replica and {thin} is created"
+    )
+)
+def volume_belonging_to_a_volume_group_with_one_replica_and_thin_is_created(thin):
+    """volume belonging to a volume group with one replica and <thin> is created."""
+    try:
+        if pytest.vol_id != "":
+            ApiClient.volumes_api().put_volume(
+                pytest.vol_id,
+                CreateVolumeBody(
+                    VolumePolicy(False),
+                    replicas=1,
+                    size=VOLUME_SIZE,
+                    thin=bool(thin),
+                    volume_group=VG_PARAM,
+                ),
+            )
+        else:
+            ApiClient.volumes_api().put_volume(
+                VG_VOLUME_UUID_1,
+                CreateVolumeBody(
+                    VolumePolicy(False),
+                    replicas=1,
+                    size=VOLUME_SIZE,
+                    thin=bool(thin),
+                    volume_group=VG_PARAM,
+                ),
+            )
+            pytest.vol_id = VG_VOLUME_UUID_2
+        pytest.exception = False
+    except openapi.exceptions.ServiceException:
+        pytest.exception = True
+
+
+@then("3 new replicas should be created")
+def new_replicas_should_be_created():
+    """3 new replicas should be created."""
+    vg_vols = get_volume_group_volumes()
+    for vol in vg_vols:
+        assert vol.spec.num_replicas == 2
+
+
+@then("3 replicas should be removed")
+def replicas_should_be_removed():
+    """3 replicas should be removed."""
+    vg_vols = get_volume_group_volumes()
+    for vol in vg_vols:
+        assert vol.spec.num_replicas == 2
+
+
+@then("the scale down operation should fail")
+def the_scale_down_operation_should_fail():
+    """the scale down operation should fail."""
+    assert (
+        "'scale down below two replicas' for 'VolumeGroup' is unsupported"
```

Comment:
not sure we should do this, seems quite brittle to compare strings

Reply:
I know, but I wanted it to fail with exactly this error. Although I check with the `Aborted` as well. Any suggestions?

Reply:
Changed it to check for the `SvcError :: RestrictedReplicaCount` and `FailedPrecondition`.